File: samples/java/ZoneWorld/Shared/src/main/java/systems/zlink/samples/zoneworld/shared/Messages.java
package systems.zlink.samples.zoneworld.shared;

import java.util.List;

public final class Messages {
    private Messages() {}

    public record PlayerView(String playerId, int x, int y, String zoneId, boolean isBot) {}

    public record NodeView(
            String nodeId,
            boolean registered,
            boolean connected,
            boolean maintenance,
            List<String> zones,
            int playerCount) {}

    public record JoinWorldMsg(String playerId) {}

    public record JoinWorldNotify(String playerId, String zoneId, int x, int y, String error) {}

    public record MoveMsg(int x, int y) {}

    public record ZoneStateNotify(String zoneId, long tick, List<PlayerView> players) {}

    public record ZoneChangedNotify(String playerId, String zoneId) {}

    public record WorldAnnounceNotify(String announcementId, String text) {}

    public record MoveRejectedNotify(String reason, int x, int y) {}

    public record WatchNodesReq() {}

    public record WatchNodesRes(List<NodeView> nodes) {}

    public record NodeStatusNotify(
            String nodeId,
            boolean registered,
            boolean connected,
            boolean maintenance,
            List<String> zones,
            int playerCount) {}

    public record NodeAlertNotify(String nodeId, String kind, String detail, String occurredAt) {}

    public record AnnounceWorldReq(String text) {}

    public record AnnounceWorldRes(String announcementId) {}

    public record SetMaintenanceReq(String nodeId, boolean enabled) {}

    public record SetMaintenanceRes(
            String nodeId, boolean enabled, List<String> zones, String error) {}

    public record NodeDiagnosticsReq(String nodeId) {}

    public record NodeDiagnosticsRes(
            String nodeId,
            List<String> zones,
            int playerCount,
            boolean maintenance,
            String error) {}

    public record RelocationPairReq() {}

    public record RelocationPairRes(
            String sourceZoneId,
            String targetZoneId,
            String sourceOwnerNodeRid,
            String targetOwnerNodeRid,
            String error) {}

    public record ActorLocationProbeReq(String actorId) {}

    public record ActorLocationProbeRes(
            String actorId, long objectGeneration, String ownerNodeRid, String error) {}

    public record FreshActorProbeReq(String actorId) {}

    public record FreshActorProbeRes(
            String actorId, long objectGeneration, String ownerNodeRid, String error) {}

    public record WorldAnnounceEvent(String announcementId, String text) {}

    public record NodeMaintenanceChangedEvent(String nodeId, boolean enabled) {}

    public record DeliverAnnounceMsg(String announcementId, String text) {}

    public record BotTickMsg() {}

    public record EnterWorldReq(int x, int y, boolean isBot, int dirX, int dirY) {}

    public record EnterWorldRes(String zoneId, int x, int y, String error) {}

    public record ReportSpotEventMsg(
            String nodeId, String kind, String detail, String occurredAt) {}

    public record ReportNodeStatusMsg(
            String nodeId, List<String> zones, int playerCount, boolean maintenance) {}

    public record ZoneBorderEvent(
            String fromZoneId, String toZoneId, long tick, List<PlayerView> players) {}

    public record EnterZoneReq(
            String playerId,
            int x,
            int y,
            boolean isBot,
            boolean initialEntry,
            String fromZoneId,
            boolean crashBoundaryProbe) {}

    public record EnterZoneRes(String zoneId, String error) {}

    public record UpdatePositionMsg(String playerId, int x, int y, boolean isBot) {}

    public record DeliverZoneStateMsg(String zoneId, long tick, List<PlayerView> players) {}

    public record DeliverZoneChangedMsg(String playerId, String zoneId) {}

    public record DeliverWorldAnnounceMsg(String announcementId, String text) {}

    public record MessageFollowProbeReq(String actorId, String probeId, byte[] payload) {}

    public record MessageFollowProbeRes(String probeId, byte[] payload) {}

    public record MessageFollowProbeMsg(String actorId, String probeId, byte[] payload) {}

    public record CrashRelocationProbeMsg(int x, int y) {}

    public record CrashRelocationProbeRes(String error) {}
}

File: samples/kotlin/ZoneWorld/Shared/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/shared/Messages.kt
package systems.zlink.samples.kotlin.zoneworld.shared

object Messages {
    data class PlayerView(
        val playerId: String,
        val x: Int,
        val y: Int,
        val zoneId: String,
        val isBot: Boolean,
    )

    data class NodeView(
        val nodeId: String,
        val registered: Boolean,
        val connected: Boolean,
        val maintenance: Boolean,
        val zones: List<String>,
        val playerCount: Int,
    )

    data class JoinWorldMsg(val playerId: String)

    data class JoinWorldNotify(
        val playerId: String,
        val zoneId: String,
        val x: Int,
        val y: Int,
        val error: String? = null,
    )

    data class MoveMsg(val x: Int, val y: Int)

    data class ZoneStateNotify(val zoneId: String, val tick: Long, val players: List<PlayerView>)

    data class ZoneChangedNotify(val playerId: String, val zoneId: String)

    data class WorldAnnounceNotify(val announcementId: String, val text: String)

    data class MoveRejectedNotify(val reason: String, val x: Int, val y: Int)

    class WatchNodesReq

    data class WatchNodesRes(val nodes: List<NodeView>)

    data class NodeStatusNotify(
        val nodeId: String,
        val registered: Boolean,
        val connected: Boolean,
        val maintenance: Boolean,
        val zones: List<String>,
        val playerCount: Int,
    )

    data class NodeAlertNotify(
        val nodeId: String,
        val kind: String,
        val detail: String,
        val occurredAt: String,
    )

    data class AnnounceWorldReq(val text: String)

    data class AnnounceWorldRes(val announcementId: String)

    data class SetMaintenanceReq(val nodeId: String, val enabled: Boolean)

    data class SetMaintenanceRes(
        val nodeId: String,
        val enabled: Boolean,
        val zones: List<String>,
        val error: String? = null,
    )

    data class NodeDiagnosticsReq(val nodeId: String)

    data class NodeDiagnosticsRes(
        val nodeId: String,
        val zones: List<String>,
        val playerCount: Int,
        val maintenance: Boolean,
        val error: String? = null,
    )

    class RelocationPairReq

    data class RelocationPairRes(
        val sourceZoneId: String,
        val targetZoneId: String,
        val sourceOwnerNodeRid: String,
        val targetOwnerNodeRid: String,
        val error: String? = null,
    )

    data class ActorLocationProbeReq(val actorId: String)

    data class ActorLocationProbeRes(
        val actorId: String,
        val objectGeneration: Long,
        val ownerNodeRid: String,
        val error: String? = null,
    )

    data class FreshActorProbeReq(val actorId: String)

    data class FreshActorProbeRes(
        val actorId: String,
        val objectGeneration: Long,
        val ownerNodeRid: String,
        val error: String? = null,
    )

    data class WorldAnnounceEvent(val announcementId: String, val text: String)

    data class NodeMaintenanceChangedEvent(val nodeId: String, val enabled: Boolean)

    data class DeliverAnnounceMsg(val announcementId: String, val text: String)

    class BotTickMsg

    data class EnterWorldReq(
        val x: Int,
        val y: Int,
        val isBot: Boolean,
        val dirX: Int,
        val dirY: Int,
    )

    data class EnterWorldRes(val zoneId: String, val x: Int, val y: Int, val error: String? = null)

    data class ReportSpotEventMsg(
        val nodeId: String,
        val kind: String,
        val detail: String,
        val occurredAt: String,
    )

    data class ReportNodeStatusMsg(
        val nodeId: String,
        val zones: List<String>,
        val playerCount: Int,
        val maintenance: Boolean,
    )

    data class ZoneBorderEvent(
        val fromZoneId: String,
        val toZoneId: String,
        val tick: Long,
        val players: List<PlayerView>,
    )

    data class EnterZoneReq(
        val playerId: String,
        val x: Int,
        val y: Int,
        val isBot: Boolean,
        val initialEntry: Boolean,
        val fromZoneId: String,
        val crashBoundaryProbe: Boolean,
    )

    data class EnterZoneRes(val zoneId: String, val error: String? = null)

    data class UpdatePositionMsg(val playerId: String, val x: Int, val y: Int, val isBot: Boolean)

    data class DeliverZoneStateMsg(
        val zoneId: String,
        val tick: Long,
        val players: List<PlayerView>,
    )

    data class DeliverZoneChangedMsg(val playerId: String, val zoneId: String)

    data class DeliverWorldAnnounceMsg(val announcementId: String, val text: String)

    data class MessageFollowProbeReq(
        val actorId: String,
        val probeId: String,
        val payload: ByteArray,
    )

    data class MessageFollowProbeRes(val probeId: String, val payload: ByteArray)

    data class MessageFollowProbeMsg(
        val actorId: String,
        val probeId: String,
        val payload: ByteArray,
    )

    data class CrashRelocationProbeMsg(val x: Int, val y: Int)

    data class CrashRelocationProbeRes(val error: String? = null)
}
